---
title: Regional Patterns of Gene Expression in Human and Chimpanzee Brains
author: "Lihua Wu"
date: "Jan 14, 2020"
output: html_document
---


In this project, I did an analysis of gene expression in various regions of primate brains.

I first normalize the data, which came from a reference design using a two-channel microarray.

```{r}
require(hexbin)
require(statmod)
require(limma)
load("ApoAI.RData")
ls()

class(RG)
names(RG)
RG$printer
names(RG$genes)
names(RG$targets)

##The *is.finite* command checks for finite numbers.  I first save the log2 data and then replace the negative infinities by 8, which is just a bit smaller than the smallest of the valid numbers####
lR=log2(RG$R)
lR[!is.finite(lR)]=8

###do the *hexplot plots for R*##
par(mfrow=c(1,1))
plot(hexplom(lR[,1:4]))

#do the *hexplot plots for G*##
lG=log2(RG$G)
lG[!is.finite(lG)]=8
par(mfrow=c(1,1))
plot(hexplom(lG[,1:4]))
```

There are some worrisome features on these plots.  Next I use the *plotMA* command to get MA plots for the first 4 microarrays.

```{r}
par(mfrow=c(2,2))
for (i in 1:4) limma::plotMA(RG[,i])
```

```{r}
MA=normalizeWithinArrays(RG)
```

```{r}
par(mfrow=c(2,2))
for (i in 1:4) limma::plotMA(MA[,i])
```

```{r}
par(mfrow=c(1,1))
boxplot(MA$M[,1:4])
```

```{r}
design <- matrix(c(rep(1,16), rep(0,8),rep(1,8)),ncol=2)#set up the design matrix
colnames(design) <-c("WT-Ref","KO-WT")
design
```

```{r}
#Fitting a linear model
fit <- lmFit(MA,design=design)
names(fit)
```
```{r}
#Empirical Bayes statistics
fit <- eBayes(fit)
names(fit)
```

```{r}
# Display tables of differentially expressed genes
topTable(fit,coef="KO-WT")
numGenes <- nrow(RG$genes)
completeTableKOvsWT <- topTable(fit,coef="KO-WT",number=numGenes)
write.table(completeTableKOvsWT, file="KOvsWTgenes.xls", sep="\t",quote=FALSE,col.names=NA)
```

```{r}
#Draw MA plot using coefficients fitted from the linear model
M <- fit$coefficients[,"KO-WT"]
A <- fit$Amean
ord <- order(fit$lods[,"KO-WT"],decreasing=TRUE)
top10 <- ord[1:10]
plot(A,M,pch=16,cex=0.1)
text(A[top10],M[top10],labels=substring(fit$genes[top10,"NAME"],1,5),cex=0.8,col="blue")
```

I then do the differential expression analysis.  This is an example of the two-sample t-test.  I can readily set up the linear model for this in *limma*.  I can either use a single indicator variable, in which case our t-test will be the test of whether the regression slope (coefficient 2) is zero, or I can use the cell means model followed by a contrast.

```{r}
design <- matrix(c(rep(1,16), rep(0,8),rep(1,8)),ncol=2)#set up the design matrix
colnames(design) <-c("WT-Ref","KO-WT")
design
```

```{r echo=TRUE}
#compare smaple variance of control WT and KO for every gene.
varWT <- var(lR[,1:8]) # Sample variance of the control samples. 
varKO <- var(lR[,9:16]) # Sample variance of the knockout samples. 
hist(varWT)
hist(varKO)
deltaVar <- varWT - varKO # The differences between the sample variances of # both control and knockout groups. 
hist(deltaVar) 
```

# Differential Expression Analysis of Affymetrix microarrays

```{r}
require(hexbin)
require(statmod)
require(affy)
require(limma)
```

```{r}
qval=function(p){
  m=length(p)
  ord=order(p)
  pi0=min(2*sum(p>0.5)/m,1)
  q=pi0*m*p/rank(p)
  q[ord[m]]=min(q[ord[m]],1)
  for (i in (m-1):1){
  q[ord[i]]=min(q[ord[i]],q[ord[i+1]],1)
}
list(pi0=pi0,qval=q)
}
```

#The Data
Khaitovich et al (2004) considered gene expression in 7 homologous regions of human and chimpanzee brains. There were 3 human brains and 3 chimpanzee
brains available for the study. Each brain was dissected to obtain tissue samples for each of the 7 regions.  This is called a split plot design.

In a split plot design, the subplot factor - in this case the brain region - is arranged in randomized complete block design.  The individual brains are the blocks.  However, the blocks themselves also have a factor - in this case species.  The interaction between species and region is also considered a subplot effect.  In an ANOVA with a single response, we would have different estimate of the SD when testing hypotheses about the whole plot and subplot errors.  *limma* gets around this by estimating the ratio of these two errors (the intra-class correlation) and then using the correlation to estimate the appropriate standard errors for the contrasts.

The factors "species" and "region" are arranged in a balanced factorial design, because each combination of species and region was sampled with the
same number of biological replicates.  However, there is also a blocking factor ``brain'' with 6 levels representing the 6 individuals.
 
 
 
The samples were hybridized to a variety of Affymetrix@ Genechips and are available as experiment E-AFMX-2 at
http://www.ebi.ac.uk//aerep//dataselection//?expid=352682122.  We will use only 4 of the brain regions: prefrontal cortex, caudate nucleus, cerebellum and
Broca's region and only one of the Genechips, hgu95b with one hybridization per sample.

Read the cel files into an AffyBatch object called **brainBatch**. 

To start, I need to load the data into R and do some basic quality assessment.  The *ReadAffy* command finds all the CEL files and reads them in an object called an *AffyBatch*.   

```{r}
celPath="D:/study/STAT555 statistics in genomics/STAT555 statistics in genomics/week7/brain"
brainBatch=ReadAffy(celfile.path=celPath)
```

The sample names for *brainBatch* are the CEL file names, which are not informative.  We will replace them with more informative names, and then 
extract the probewise raw expression values for quality assessment.  The *paste* and *rep* command are very handy for creating 
names.    The microarray names are coded 
``a_xny" where n is the replicate number,
x is either ``c" for chimpanzee or ``h" for human, and the brain regions are a) prefrontal cortex, d) caudate nucleus e) cerebellum or f) Broca's region.
First print the sampleNames to be sure the microarrays are in the right order.  Then replace the names with the more informative names. 

```{r}
# Print the samplenames and replace the names with more informative ones. 
sampleNames(brainBatch)
sampleNames(brainBatch)=
  paste(rep(c("CH","HU"),each=12),rep(c(1:3,1:3),each=4),
  rep(c("Prefrontal","Caudate","Cerebellum","Broca"),6),sep="")
```

Display *brainBatch* to get a sense of what is in it.  If you type the name, R will go to the Bioconductor website to try to download the annotation.  If it is successful you will discover that these are *hgu95av2* microarrays.  You could go to the Affymetrix website to learn more about these microarrays.  However, everything you need for the basic analysis was already downloaded!  This includes which probes are in each probeset, and the location of the probes in the sample datafile.  Notice that there are 640\*640=409600 individual probes, but only 12625 probesets representing genes.

```{r}
# Display *brainBatch*.
brainBatch
```

If you have trouble with this, you might need to go to the Bioconductor website to download the ``cdf" file for *hgu95av2* microarrays.  This information is required by the software so that it knows which probes belong in each probeset.

We should at minimum check quality by doing some scatterplot matrices of the log2(expression) values.    An *AffyBatch* is a special type of *expressionSet* object, so all the commands used for *expressionSet*s are available.  In particular, the *exprs* command will extract the 409600 probe by 24 microarray of probewise expression values.  To look at 409600 expression values, we will need to use *hexplom* again. There are a lot of samples to check.    It is convenient to do them by brain region, as this would give 6 plots at a time with gene expression we might expect to be roughly comparable (at least within species.)   

```{r}
# extract the 409600 probe by 24 microarray of probewise expression values.
brain.exprs <- exprs(brainBatch)
```

```{r}
# Plot the scatter plot of Prefrontal region
plot(hexplom(log2(brain.exprs[,paste(rep(c("CH","HU"),each=3),c(1:3,1:3),rep("Prefrontal",6),sep="")])))

# Plot the scatter plot of Caudate region
plot(hexplom(log2(brain.exprs[,paste(rep(c("CH","HU"),each=3),c(1:3,1:3),rep("Caudate",6),sep="")])))

# Plot the scatter plot of Cerebellum region
plot(hexplom(log2(brain.exprs[,paste(rep(c("CH","HU"),each=3),c(1:3,1:3),rep("Cerebellum",6),sep="")])))

# Plot the scatter plot of Broca region
plot(hexplom(log2(brain.exprs[,paste(rep(c("CH","HU"),each=3),c(1:3,1:3),rep("Broca",6),sep="")])))
```

I set up the treatment names and blocks.  This is readily done using *paste* and *rep*.  The treatment names are the same as
the sample names, but the replicate numbers are dropped.
There is one block label for each brain.

Create a variable *blocks* which has labels for each subject.  This will need to be in the same order as the samples. For example, each of the 4 brain samples for Human subject 1 should be labeled HU1. Print out the values.

```{r}
# Set up a variable which has labels for each subject.
blocks <- factor(rep(1:6,each=4))
print(blocks)
```

Create a variable *trts* which has labels for each combination of species and brain region.  For example, each of the 3 human cerebellum samples should be labeled HUCerebellum. Print out the values.

```{r}
# Create a variable which has labels for each combination 
# of species and brain region.
trts <- factor(paste(rep(c("CH","HU"),each=12), rep(c("Prefrontal","Caudate","Cerebellum","Broca"),6), sep=""))
print(trts)
```

The next step is to normalize the expression values and combine into probeset expression summaries using RMA.  This can be done authomatically by applying the *rma* function directly to the *AffyBatch* (not the expression values).  Note that the *AffyBatch* stores the probeset information while the expression values are only the intensity of each probe. The output of *rma* is also an expressionSet.  Name it *brain.rma*.


```{r}
# Normalize the expression values using RMA
brain.rma <- rma(brainBatch)
```

```{r}
# Display brain.rma.
brain.rma
```

Redo the *hexplom* plots. (Note that the *rma* command creates an *expressionSet*.  You need to use the *exprs* command to get the actual expression values.) Are there any obvious problems with the data?
 

```{r}
brain.exprs.rma <- exprs(brain.rma)
```

```{r}
# Redraw the *hexplom* plot of Prefrontal region.
plot(hexplom(brain.exprs.rma[,paste(rep(c("CH","HU"),each=3),c(1:3,1:3),rep("Prefrontal",6),sep="")]))

# Plot the scatter plot of Caudate region.
plot(hexplom(brain.exprs.rma[,paste(rep(c("CH","HU"),each=3),c(1:3,1:3),rep("Caudate",6),sep="")]))

# Plot the scatter plot of Cerebellum region.
plot(hexplom(brain.exprs.rma[,paste(rep(c("CH","HU"),each=3),c(1:3,1:3),rep("Cerebellum",6),sep="")]))

# Plot the scatter plot of Broca region.
plot(hexplom(brain.exprs.rma[,paste(rep(c("CH","HU"),each=3),c(1:3,1:3),rep("Broca",6),sep="")]))
```

We are now ready to perform analysis in *limma*. The steps are:

* Compute the correlation among regions from the same brain (averaged across all the genes and all the brains.)
* Compute \(S^2_p\) the pooled variance for each gene.  
* Create the coefficient matrix for the contrasts.
* Compute the estimated contrasts.
* Compute the moderated contrast t-test for each gene.
* Plot the histogram of p-values for each contrast for each gene.
* Create the list of significant genes based on the p-values, adjusted p-values or FDR estimates.

We will go through these steps below. 

## Compute \(S^2_p\)

There are 3 steps to computing the pooled variance.

* Create a design matrix for the treatment effects.
* If there are blocks, compute the within block correlation for each gene.
* Fit the model for the treatment effects to obtain the pooled variance.

A design matrix is a matrix whose columns are the indicator variables for the effects. There is one row for each sample. We are going to set up the cell means model, so we will have an indicator variable for each brain region in each species.  We have already set up the *trts* variable to help us with this.  This is where the idea of *factors* in R comes in handy.  When R sees an equation with a factor (or a character vector) it knows that you want to create an indicator variable for each level of the factor.

The *model.matrix* command sets up the design matrix.  By default there is a column of 1's for the intercept.  In the cell means model there is no intercept.  We indicate this by putting "0+" as the leading part of our model formula.

```{r}
design.trt=model.matrix(~0+trts)
print(design.trt)
```


```{r}
# Compute the correlation of genes within the blocks. 
corfit <- duplicateCorrelation(brain.rma, design = design.trt, block = blocks)
```


```{r}
# Plot the histogram of intraclass correlation
hist(tanh(corfit$atanh.correlations))

# Print out the consensus correlation.
print(corfit$consensus.correlation)
```

## Pooled within treatment variance

We are now ready to compute the pooled sample variance for each gene.  
As a side effect, we also compute the sample mean expression of each gene in each treatment (remembering that
after RMA normalization, the data are on the log2 scale).
We do this by fitting the linear model, giving the treatments (for estimating the means), the blocking factor (to tell *limma* which observations should be correlated) and the consensus correlation.

```{r}
fitTrtMean <- lmFit(brain.rma, design = design.trt, block = blocks, 
                    cor = corfit$consensus.correlation)
```

The output *fitTrtMean* has several components, but only 2 of these are of interest.  Component *coefficients* contains the mean expression
for each gene in each treatment.  Component *sigma* has the estimate of \(S_p\).  (Notice this the pooled SD, not the pooled variance.)


```{r}
hist(fitTrtMean$coefficients[,1])
```


```{r}
hist((fitTrtMean$sigma)^2,breaks=100) 
```

## Create the coefficient matrix for the contrasts

We need to compute the coefficient matrix for any contrasts we want to do.  We will obtain
the pooled variances from *fitTrtMean*.

We need to decide what contrasts are interesting to us.  For this lab, we will look at 3 contrasts: 

*  Average chimpanzee versus average human
*  Cerebellum versus Broca's region averaged over species.
*  The interaction between species and the comparison of cerebellum to Broca's region.

Note that the treatment names are taken from the columns of the design matrix.  To make more useful names for the final output, we will want to rename the
columns of the contrast matrix.

a)nple setting up the contrast of the average Chimp versus human effect, and the interaction between species and the two brain regions, cerebellum and Broca region.  Set up a command to create the contrast matrix for all the desired contrasts.  

```{r}
contrast.matrix=makeContrasts(
  ChVsHu=(trtsCHBroca+trtsCHCaudate+trtsCHCerebellum+trtsCHPrefrontal)/4
     -(trtsHUBroca+trtsHUCaudate+trtsHUCerebellum+trtsHUPrefrontal)/4, # Contrast of the average Chimp versus human effect.
  CeVsBr=(trtsCHCerebellum+trtsHUCerebellum)/2-(trtsCHBroca+trtsHUBroca)/2, # Contrast of Cerebellum versus Broca's region averaged over species
  Interact=(trtsCHCerebellum-trtsCHBroca)-(trtsHUCerebellum-trtsHUBroca), # Contrast of the interaction between species and the comparison of cerebellum to Broca's region.
  levels=design.trt)
print(contrast.matrix) 
```

The resulting contrast coefficient matrix has one row for each treatment and one column for each contrast. Check this by printing the matrix.

## Compute the estimated contrasts and compute the moderated t-tests for each contrast.

We simply fit the contrast matrix to the previous fitted model:

```{r}
# Fit the three contrast matrices to the previous fitted model.
fit.contrast=contrasts.fit(fitTrtMean,contrast.matrix)
```

The *eBayes* command will compute the consensus pooled variance, and then use it to compute the empirical Bayes (moderated) pooled variance 
for each gene.  This also adjusts the degrees of freedom for the contrast t-tests.  The command also computes the t-tests and associated p-values.

```{r}
# Empirical Bayes moderated t-tests.
efit.contrast=eBayes(fit.contrast)
```

The interesting components of this output are the moderated variances, *s2.post*, the estimated contrasts, which are stored in the component *coefficient* and the contrast 
p-values,
which are stored in component *p.value*.

The pooled sample standard deviations are stored in component *sigma*.  Plot the sample variances (x-axis, \(S_p^2\) versus the moderated variances (y-axis, *s2.post*). Use the *abline* command to add the line y=x.  I notice that most of the moderated variances are smaller than sample variances. There are 6823 moderated variances smaller than the sample variances. 

```{r}
# Plot the sample variance versus the moderated variances.
plot(efit.contrast$s2.post ~ (efit.contrast$sigma)^2, xlab="sample variances", ylab="moderated variances")
abline(a=0, b=1,col="red") # Add the line y=x.
print(sum(efit.contrast$s2.post < (efit.contrast$sigma)^2)) # Compute the number of moderated variances which are larger than the sample variances.
```

I draw the histogram of the estimated value of the contrast coefficient for the contrast between cerebellum and Broca's region. This histogram shows that major group of genes have very small difference in gene expression in the two brain regions (the contrast of most genes are located around 0). 

```{r}
# Draw the histogram of the estimated coefficients for the contrast of two brain regions.
hist(efit.contrast$coefficient[,2])
```

I get the largest absolute difference. the "33925_at" probeset has the largest difference in estimated coefficient. The estimated fold change for this gene is 5.028054. It is down regulated in cerebellum compared to Broca's region. 

```{r}
sort.coefficient <- sort(abs(efit.contrast$coefficient[,2])) # Sort the value from min to max. 
sort.coefficient[12625] # Largest absolute value of coefficient
min <- min(efit.contrast$coefficient[,2]) # Minimum value of estimated coefficient
max <- max(efit.contrast$coefficient[,2]) # Maximum value of estimated coefficient
print(min)
print(max) # Check the largest absolute difference is up or down regulated. 
```

The annotation of this probeset is search for online. The gene name is neurogranin (protein kinase C substrate, RC3).  The p.value for the probeset with largest estimated coefficient is 9.966758e-23. 

```{r}
row.names <- rownames(efit.contrast$p.value)
for (i in 1:12625) {
        if (row.names[i] == "33925_at") {
                print(efit.contrast$p.value[,2][i])
        }
}
```

I draw the histogram of the estimated interaction between species and brain regions cerebellum and Broca region. The frequency of probesets with near 0 difference is slightly higher than the second contrast. The largest estimate of this contrast is 3.410219, and the probeset name is "38285_at". This indicates smaller number of genes have large fold changes between the two brain regions in individual species. Genes are less differentially expressed between regions in the same species.  

```{r}
Isort.coeff <- sort(abs(efit.contrast$coefficient[,3])) # Sort the value from min to max. 
Isort.coeff[12625] # Largest absolute value of coefficient
min <- min(efit.contrast$coefficient[,3]) # Minimum value of estimated coefficient
max <- max(efit.contrast$coefficient[,3]) # Maximum value of estimated coefficient
print(min)
print(max) # Check the largest absolute difference is up or down regulated. 
```

I draw the histograms of p.values for each of the contrasts. The second histogram has the appropriate shape for FDR estimation, while the first and third histogram do not show a typical non-null distribution. Even though they have peak close to 0, they have huge false discovery genes.    

```{r}
hist(efit.contrast$p.value[,1], main="CHVsHU.contrast")
hist(efit.contrast$p.value[,2],main="CVsBr.contrast")
hist(efit.contrast$p.value[,3], main="Interact.contrast")
```


##Adjust the p-values.##

All of the comparisons have a large percentage of differentially expressing genes.  We will want to use
a multiple comparisons procedure that adapts to having a large number of non-null hypotheses such as the Storey method.

*limma* has a number of methods for adjusting the p-values and/or estimating FDR for each p-value.  However, I like Storey's q-value method, because it estimates \(\pi_0\) the percentage of non-differentially expressing genes and uses the estimate in computing the q-value.  We will use the function we created to do this.

I estimate the q-values for each contrast. For CHvsHU, \(\pi_0\) is 0.7767129, and the number of probeset with q < 0.05 is 903. For CVsBr,\(\pi_0\) is 0.4250297, and the number of probeset with q < 0.05 is 6005. For Interact contrast,\(\pi_0\) is 0.9156436, and the number of probeset with q < 0.05 is 212.

```{r}
CHvsHU.qvalue <- qval(efit.contrast$p.value[,1]) # Get q-values for the efit.contrast.
print(CHvsHU.qvalue$pi0)
print(CHvsHU.qvalue$qval[1:20])
i <- NULL
j <- 0 
for (i in 1:12625) {
        if (CHvsHU.qvalue$qval[i] < 0.05) {
          j=j+1      
        }
}
print(j) # Print number of probeset with q<0.05. 
```

```{r}
CvsBr.qvalue <- qval(efit.contrast$p.value[,2]) # Get q-values for the CvsBr contrast.
print(CvsBr.qvalue$pi0)
print(CvsBr.qvalue$qval[1:20])

a <- NULL
b <- 0
for (a in 1:12625) {
        if (CvsBr.qvalue$qval[a] < 0.05) {
                b=b+1
        }
}
print (b) # Print number of probeset with q<0.05. 
```

```{r}
Interact.qvalue <-qval(efit.contrast$p.value[,3])# Get q-values for the Interact contrast.
print(Interact.qvalue$pi0)
print(Interact.qvalue$qval[1:20])
x <- NULL
y <- 0
for (x in 1:12625) {
        if (Interact.qvalue$qval[x] < 0.05) {
                y=y+1
        }
}
print (y) # Print number of probeset with q<0.05.
```

The second contrast, the contrast of cerebellum to Broca's region has the most differential expression ( It has the largest number of probesets with q < 0.05. This is what I would expect. When I look at the histograms of p.value s of these three contrasts, the histogram of the contrast of the two brain regions is nicely shaped, and the frequency of p.value < 0.05 is very high.       


#Compute the gene list

The most statistically significant genes for each contrast can be assembled into spreadsheets.  There are several ways to do this.
*limma* provides 2 functions, *topTable* and *decideTests* to assemble gene lists.  I prefer to compute FDR or q-value estimates or adjusted p-values for 
each gene and output the treatment means and estimated contrasts, p-values and FDR or q-values to a comma separated text file which I can import 
to a spreadsheet.

To use  *topTable*, select a contrast and one of the adjustment methods.  Of those available, Benjamini and Yuketiel (2001) (``BY") 
is a good general purpose choice.  You also need probeset ids, which can either be extracted from the original data or from the row names of the
p-value matrix.

To limit the output to the most statistically significant genes, set the input parameter *p.value* to the 
maximum adjusted p-value or estimated FDR 
that you want to consider and the input parameter *n* to the maximum number of genes you want on the list.  If you want a complete list,
set *p.value=1.0* and *n=X* where X is bigger than the total number of probesets on the microarray.

For example to get the top 10 genes with "BY adjusted" \(p<10^{-5}\) for the overall species comparison and for the interaction contrast:

```{r}
genes=geneNames(brainBatch) # Extract the names of the probesets. 
CHvsHU <- topTable(efit.contrast,coef=1,adjust.method="BY",n=10,p.value=1e-5,genelist=genes)# Get the top 10 genes for the overall species comparison.
CvsBr <- topTable(efit.contrast,coef=2,adjust.method="BY",n=10,p.value=1e-5,genelist=genes)# Get the top 10 genes for the two brain regions comparison.
Interact <- topTable(efit.contrast,coef=3,adjust.method="BY",n=10,p.value=1e-5,genelist=genes)# Get the top 10 genes for the species and the brain region comparison.
CHvsHU[1,]
CvsBr[1,]
Interact[1,]
```

The columns of the table are the row number of the gene, the gene id, the estimated contrast, the expression mean over all microarrays, contrast t-value, 
contrast p-value, contrast adjusted p-value or estimated FDR and the estimated log-odds probability ratio that the gene is differentially expressed.

The *decideTests* function can be used to create indicator variables for significance of contrasts with a variety of options.

As an alternative, *write.table* can be used to create a comma separated text file, using *cbind* to concatenate matrices.Append adjusted p-values to the table above using either *p.adjust* or *qval*.

```{r}
# Use p.adjust with the BH method to get adjusted P.values of these 3 contrasts, 
# and create a comma seperated text file. 
adjP1 <- p.adjust(efit.contrast$p.value[,1],method="BH", n= length(efit.contrast$p.value))
adjP2 <- p.adjust(efit.contrast$p.value[,2],method="BH",n= length(efit.contrast$p.value))
adjP3 <- p.adjust(efit.contrast$p.value[,3],method="BH",n= length(efit.contrast$p.value))
adjP <- cbind(adjP1,adjP2,adjP3)
write.table(file="myresult.txt",cbind(genes,fitTrtMean$coefficients,efit.contrast$coefficients,efit.contrast$p.value,adjP[,1:3]),row.names=F,col.names=c("GeneID",colnames(fitTrtMean$coefficients),colnames(efit.contrast$p.value),paste("p",colnames(efit.contrast$coefficients)),paste("adjp",colnames(efit.contrast$p.value))),sep=",")

# Read the file myresult.txt into R and print the first row.
myresult <- read.csv("D:/study/STAT555 statistics in genomics/STAT555 statistics in genomics/week7/myresult.txt", header=TRUE, sep =",", dec=".", na.strings="NA")
myresult[1,]
```

```{r}
toLatex(sessionInfo())
print(gc())
```

Benjamini, Y., and Yekutieli, D. (2001). The control of the false discovery rate in multiple testing under dependency. 
*Annals of Statistics*,**29**: 1165-1188.

Khaitovich, P., Muetzel, B., She, X., Lachmann, M., Hellmann, I., Dietzsch, J., Steigele, S., Do, H. H., Weiss, G., 
 Enard, W., Heissig, F., Arendt, T., Nieselt-Struwe, K., Eichler, E. E., Paabo, S. (2004)
Regional patterns of gene expression in human and chimpanzee brains.
*Genome Research*, **14** (8) :1462--73.

Smyth, G. K. (2004). Linear models and empirical Bayes methods for assessing differential expression in microarray experiments. 
*Statistical Applications in Genetics and Molecular Biology*, **3**, Article 3. http://www.bepress.com/sagmb/vol3/iss1/art3.

Storey JD. (2003) The positive false discovery rate: A Bayesian interpretation and the q-value. *Annals of Statistics*, **31**: 2013--2035. 